crates/engine/tree/src/tree/payload_validator.rs

```diff
@@ -60,13 +60,38 @@ pub struct TreeCtx<'a, N: NodePrimitives> {
     state: &'a mut EngineApiTreeState<N>,
     /// Reference to the canonical in-memory state
     canonical_in_memory_state: &'a CanonicalInMemoryState<N>,
+    /// Optional precomputed state provider and builder to avoid redundant lookups.
+    /// This is set by [`crate::tree::EngineApiTreeHandler`] after validating parent state exists.
+    /// The builder is type-erased via `dyn Any` because adding a generic `P` to `TreeCtx` would
+    /// require cascading changes to the public `EngineValidator` trait signature.
+    precomputed_state: Option<PrecomputedState>,
+}
+
+/// Precomputed state provider and builder for block validation.
+///
+/// Stored as type-erased builder to avoid adding generics to `TreeCtx`.
+pub struct PrecomputedState {
```

Comment:
another nit: calling this "state" is a bit confusing, it's not carrying state but instead the StateProvider and Builder. Maybe keep it obvious and go with "StateProviderAndBuilder" ?